### PR TITLE
Add basic logger

### PR DESCRIPTION
**Pull request by SanttuRantanen**
Please review.

### CODE REVIEW

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 1]
```diff
@@ -1,18 +1,152 @@
-#ifndef STREAMER_UTILS_LOGGER_LOGGER_HPP
-#define STREAMER_UTILS_LOGGER_LOGGER_HPP
+#ifndef STREAMER_UTILS_LOGGER_LOGGER_H
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: STREAMR_LOGGER_LOGGER_HPP

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 2]
```diff
@@ -1,18 +1,152 @@
...
+#include <folly/logging/LogConfig.h>
+#include <folly/logging/LogLevel.h>
+#include <folly/logging/LogMessage.h>
+#include <folly/logging/LogWriter.h>
+#include <folly/logging/LoggerDB.h>
+#include <folly/logging/xlog.h>
+#include "StreamrWriterFactory.hpp"
+#include "StreamrHandlerFactory.hpp"
+
+namespace streamr {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: combine to streamr::logger

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 3]
```diff
@@ -1,18 +1,152 @@
...
+namespace logger {
+
+namespace {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: Do not use unnamed namespaces in header files, it is potentially dangerous, see discussion: https://stackoverflow.com/questions/357564/are-there-any-uses-for-unnamed-namespaces-in-header-files

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 4]
```diff
@@ -1,18 +1,152 @@
...
+
+static const auto envLogLevelName = "LOG_LEVEL";
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: Maybe could use constexpr std::string_view STRING_CONSTANT = "String"; for string constants instead, see discussion in https://www.reddit.com/r/cpp/comments/cw35kk/best_practices_for_efficient_string_constants/

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 5]
```diff
@@ -1,18 +1,152 @@
...
+        }
+    }
+    return std::nullopt;
+}
+
+}; // namespace
 
-namespace streamr::logger {
 class Logger {
 public:
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp:26]: Put private section first as is the Streamr style in typescript

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 6]
```diff
@@ -1,18 +1,152 @@
...
-    Logger() = default;
 
-    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
-    void log(const std::string& message) const {
-        std::cout << message << std::endl;
+    Logger(std::shared_ptr<folly::LogWriter> logWriter = nullptr)
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: Make an overloaded version Logger() so it looks nicer in IDEs? 
Also support the parameters used in the typescript version. Maybe std::optional<something> could be also useful? Anyway, see what looks nicest in auto-complete

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 7]
```diff
@@ -1,18 +1,152 @@
...
+            writerFactory_ = StreamrWriterFactory(logWriter);
+        } else {
+            writerFactory_ = StreamrWriterFactory();
+        }
+        logHandlerFactory_ = {
+            std::make_unique<StreamrHandlerFactory>(&writerFactory_)};
+        this->initializeLoggerDB(folly::LogLevel::DBG);
+    }
+
+    static Logger& get(std::shared_ptr<folly::LogWriter> logWriter = nullptr) {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: Is this class meant to be singleton or why is this needed? How many copies of the class can you create? When are they deleted? Is there a risk for memory leaks?

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 8]
```diff
@@ -1,18 +1,152 @@
...
+        static Logger instance(logWriter);
+        return instance;
+    }
+
+    void log(StreamrLogLevel streamrLevel, const std::string& message) {
+        auto follyRootLogLevel = getFollyLogLevelFromEnv();
+        auto pair = fromStreamrToFollyLogLevelMap.find(streamrLevel);
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: do not access the map directly, add a helper function

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 9]
```diff
@@ -1,18 +1,152 @@
...
+        if (pair == fromStreamrToFollyLogLevelMap.end()) {
+            // Not found from map, return
+            return;
+        }
+        if (follyRootLogLevel) {
+            if (follyRootLogLevel != loggerDB_.getCategory("")->getLevel()) {
+                // loggerDB.setLevel("", *follyLogLevel);
+                this->initializeLoggerDB(*follyRootLogLevel, true);
+            }
+            // This complicated code is mostly copy/pasted from folly's XLOG
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: put the complicated code into a helper function

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 10]
```diff
@@ -1,18 +1,152 @@
...
+                __LINE__,
+                __func__,
+                ::folly::LogStreamProcessor::APPEND,
+                message)
+                .stream();
+        }
+    }
+
+private:
+    std::shared_ptr<folly::LogWriter> logWriter_;
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: do not use underscore in member variable names

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 11]
```diff
@@ -1,18 +1,152 @@
...
+        auto rootCategoryConfig =
+            folly::LogCategoryConfig(rootLogLevel, false, {"default"});
+        folly::LogConfig config(
+            {{"default", defaultHandlerConfig}}, {{"", rootCategoryConfig}});
+        loggerDB_.updateConfig(config);
     }
 };
 
-} // namespace streamr::logger
+#define SLOG_TRACE(msg) Logger::get().log(StreamrLogLevel::trace, msg)
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: rename these to logTrace(msg) etc.

File: packages/streamr-logger/include/streamr-logger/StreamrHandlerFactory.hpp [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    std::shared_ptr<folly::LogHandler> createHandler(
+        const Options& options) override {
+        StreamrLogFormatterFactory formatterFactory;
+        return folly::StandardLogHandlerFactory::createHandler(
+            getType(), writerFactory_, &formatterFactory, options);
+    }
+
+private:
+    StreamrWriterFactory* writerFactory_;
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrHandlerFactory.hpp]: no underscore, also should a smart pointer be used here?

File: packages/streamr-logger/include/streamr-logger/StreamrHandlerFactory.hpp [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+#include <folly/logging/StandardLogHandler.h>
+#include <folly/logging/StandardLogHandlerFactory.h>
+#include <folly/logging/StreamHandlerFactory.h>
+#include "StreamrLogFormatterFactory.hpp"
+#include "StreamrWriterFactory.hpp"
+
+namespace streamr::logger {
+
+class StreamrHandlerFactory : public folly::StreamHandlerFactory {
+public:
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrHandlerFactory.hpp]: private section first

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
+#ifndef STREAMER_LOGGER_STREAMRLOGFORMATTER_H
+#define STREAMER_LOGGER_STREAMRLOGFORMATTER_H
+
+#include <iostream>
+#include <folly/Format.h>
+#include <folly/logging/LogFormatter.h>
+#include <folly/logging/LogLevel.h>
+
+namespace {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: no unnamed namespace

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+static constexpr int maxFileNameAndLineNumberLength{36};
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: let's study/discuss more what is the best practice for constants in C++26

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+    std::to_string(maxFileNameAndLineNumberLength) + "}): {}{}{}\n";
+static constexpr std::string_view truncatedFormatter =
+    "{}{}{} [{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{}] ({: <*}{}{}): {}{}{}\n";
+
+struct LogLevelData {
+    folly::StringPiece logLevelName;
+    folly::StringPiece color;
+};
+
+// FATAL cannot be used in folly because it aborts. So CRITICAL is converted
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: study typescript version to see what fatal does there, does it terminate or not?

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+// to Streamr fatal
+constexpr LogLevelData getLogLevelData(const folly::LogLevel level) {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: name function more clearly to communicate intent, what data?

Comment (ptesavol) [reply packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: add a comment about why switch statement is needed (std::map cannot be used in compile time)

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+            break;
+    }
+}
+
+} // namespace
+
+namespace streamr::logger {
+
+/*
+Format based on this Streamr log. The Filename and line number is fixed size
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: Also take the additional fields into account in the format. (context info and the extra parameter of logger.info(), warn etc )

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+        const auto timeSinceEpoch = message.timestamp.time_since_epoch();
+        const auto epochSeconds =
+            std::chrono::duration_cast<std::chrono::seconds>(timeSinceEpoch);
+        const std::chrono::milliseconds millisecs =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                timeSinceEpoch) -
+            epochSeconds;
+        time_t unixTimestamp = epochSeconds.count();
+        if (!localtime_r(&unixTimestamp, &ltime)) {
+            memset(&ltime, 0, sizeof(ltime));
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: do this in more c++ way if there is one , memset looks quite low-level c code

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+        const std::string logMessageColor = "\033[36m";
+        const std::string logMessageColorReset = "\033[0m";
+        const auto fileNameAndLineNumberLength =
+            (fileNameLength + lineNumberLength + separatorLength);
+        auto logLevelData = getLogLevelData(message.logLevel);
+        if (fileNameAndLineNumberLength <= maxFileNameAndLineNumberLength) {
+            basename = basename.toString()
+                           .append(fileNameAndLineNumberSeparator)
+                           .append(lineNumberInString);
+            auto logLine = folly::sformat(
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: add helper function to increase dryness

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+                ltime.tm_min,
+                ltime.tm_sec,
+                millisecs.count(),
+                basename,
+                logMessageColor,
+                message.logMessage,
+                logMessageColorReset);
+            return logLine;
+        } else {
+            // Truncate needed
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: use the same helper here

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatterFactory.hpp [hunk 22]
```diff
@@ -0,0 +1,24 @@
+#ifndef STREAMER_LOGGER_STANDARDLOGHANDLERFACTORY_H
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatterFactory.hpp]: STREAMER_LOGGER_STANDARDLOGHANDLERFACTORY_HPP

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp [hunk 23]
```diff
@@ -0,0 +1,175 @@
+#ifndef STREAMER_LOGGER_STREAMRLOGFORMATTER_H
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: STREAMER_LOGGER_STREAMRLOGFORMATTER_HPP

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 24]
```diff
@@ -1,18 +1,152 @@
...
+            std::make_unique<StreamrHandlerFactory>(&writerFactory_)};
+        this->initializeLoggerDB(folly::LogLevel::DBG);
+    }
+
+    static Logger& get(std::shared_ptr<folly::LogWriter> logWriter = nullptr) {
+        static Logger instance(logWriter);
+        return instance;
+    }
+
+    void log(StreamrLogLevel streamrLevel, const std::string& message) {
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: add the logging functions info() warn() error() etc to make the class interface a close to the typescript version as possibe (the macros are only extra feature). Take also into account the capturing of the file name and line number, check that they point to right places both with these new functions and macros

File: packages/streamr-logger/include/streamr-logger/Logger.hpp [hunk 25]
```diff
@@ -1,18 +1,152 @@
...
+        auto follyRootLogLevel = getFollyLogLevelFromEnv();
```
Comment (ptesavol) [inline packages/streamr-logger/include/streamr-logger/Logger.hpp]: TODO: add support for per-file log levels with env variables of the format LOG_LEVEL_MyClass.cpp=trace

**Review by ptesavol — CHANGES_REQUESTED** [review]
Please see comments

File: packages/streamr-logger/include/streamr-logger/Logger.hpp (hunk 5 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/Logger.hpp:26]: I read that the recommended way is to have public first because for the user it is important to see how to use the class or interface

File: packages/streamr-logger/include/streamr-logger/Logger.hpp (hunk 7 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/Logger.hpp]: It was the idea to use it the class without an explicit initialization. We discussed it earlier. So yes it was a singleton. So no really risk for memory leaks and not even data race because it was Meyer's Singleton. Anyway I removed it if it is not needed.

File: packages/streamr-logger/include/streamr-logger/Logger.hpp (hunk 10 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/Logger.hpp]: What is the naming convention for member variables in this project. This is the way what is used in Folly. One way is to use hungarian notation. There are many different ways.

File: packages/streamr-logger/include/streamr-logger/Logger.hpp (hunk 11 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/Logger.hpp]: I removed all macros.

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp (hunk 16 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: No it does not terminte.

File: packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp (hunk 19 above)
Comment (SanttuRantanen) [reply packages/streamr-logger/include/streamr-logger/StreamrLogFormatter.hpp]: this was copy/paste from the folly code

File: packages/streamr-logger/test/unit/StreamrLogFormatterTest.cpp [hunk 26]
```diff
@@ -0,0 +1,103 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <streamr-logger/detail/StreamrLogFormatter.hpp>
+using namespace streamr::logger::detail; // NOLINT
```
Comment (jtakalai) [inline packages/streamr-logger/test/unit/StreamrLogFormatterTest.cpp]: can linter be set to not need `// NOLINT` here? Or else add explanation why this one is exceptional.

**Review by jtakalai — APPROVED** [review]
ample docs and tests 👍

**Review by ptesavol — APPROVED** [review]
Ready to be merged
